rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+//! ```
+
+extern crate asn1;
+extern crate base64;
+extern crate openssl;
+extern crate serde_json;
+extern crate time;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::io::Write;
+use std::path::Path;
+
+use openssl::bn::BigNum;
+use openssl::bn::BigNumContext;
+use openssl::hash::MessageDigest;
+use openssl::ec;
+use openssl::nid;
+use openssl::sign::{Signer, Verifier};
+use openssl::pkey;
+
+pub struct Key(ec::EcKey);
+
+/// a Key is a helper for creating or using a VAPID EC key.
+///
+/// Vapid Keys are always Prime256v1 EC keys.
+///
+#[allow(dead_code)]
+impl Key {
+    fn new() -> Key {
+        Key(ec::EcKey::from_curve_name(Key::name()).expect(
+            "EC Prime256v1 curve not supported",
+        ))
+    }
+
+    fn name() -> nid::Nid {
+        nid::X9_62_PRIME256V1
+    }
+
+    fn group() -> ec::EcGroup {
+        ec::EcGroup::from_curve_name(Key::name()).expect("EC Prime256v1 curve not supported")
+    }
+
+    /// Read a VAPID private key stored in `path`
+    pub fn from_pem(path: String) -> Result<Key, String> {
+        let mut pem_file =
+            File::open(&Path::new(&path)).expect(&format!("Could not open file at {}", path));
+        let mut pem_data = String::new();
+        pem_file.read_to_string(&mut pem_data).expect(&format!(
+            "Could not read from file {}",
+            path
+        ));
+        let key_data = ec::EcKey::private_key_from_pem(&pem_data.into_bytes())
+            .expect("Could not read PEM key data");
+        Ok(Key(key_data))
+    }
+
+    /// Write the VAPID private key as a PEM to `path`
+    pub fn to_pem(&self, path: String) -> Result<bool, String> {
+        let key_data = self.0.private_key_to_pem().expect(
+            "Could not generate PEM for privakte key",
+        );
+        let mut pem_file =
+            File::create(&Path::new(&path)).expect(&format!("Could not open file at {}", path));
+        pem_file.write(&key_data).expect(&format!(
+            "Could not write PEM data to file {}",
+            path
+        ));
+        Ok(true)
+    }
+
+    /// Create a new Vapid key
+    pub fn generate() -> Key {
+        Key(ec::EcKey::generate(&Key::group()).expect(
+            "Could not generate EC key",
+        ))
+    }
+
+    /// Convert the private key into a base64 string
+    pub fn to_private_raw(&self) -> String {
+        // Return the private key as a raw bit array
+        let key = self.0.private_key().expect("Could not extract private key");
+        String::from_utf8(key.to_vec()).expect("Could not convert private key to string")
+    }
+
+    /// Derive a vapid key from the provided raw base64 public key string
+    fn derive_ec_public_point(bits: String) -> Key {
+        //Read a private key from a raw bit array
+        let group = Key::group();
+        let mut ctx = BigNumContext::new().unwrap();
+        let mut new_key = ec::EcKeyBuilder::new().unwrap();
+
+        let bytes: Vec<u8> = bits.into_bytes();
```

Comment:
Here the conversion from `String` to `Vec<u8>` is actually just the raw bytes of the string encoding (e.g. the ascii values), so it may need a base64 decoding step?

Reply:
Hmm. This was a somewhat troublesome function. There may be a simpler means to do this. In any respect, I should remove it since it's currently unused.